Content.Shared/DoAfter/SharedDoAfterSystem.cs

```diff
@@ -0,0 +1,385 @@
+        SubscribeLocalEvent<DoAfterComponent, ComponentGetState>(OnDoAfterGetState);
+    }
+
+    private void Add(EntityUid entity, DoAfterComponent component, DoAfter doAfter)
+    {
+        doAfter.ID = component.RunningIndex;
+        doAfter.Delay = doAfter.EventArgs.Delay;
+        component.DoAfters.Add(component.RunningIndex, doAfter);
+        EnsureComp<ActiveDoAfterComponent>(entity);
+        component.RunningIndex++;
+        Dirty(component);
+    }
+
+    private void OnDoAfterGetState(EntityUid uid, DoAfterComponent component, ref ComponentGetState args)
+    {
+        args.State = new DoAfterComponentState(component.DoAfters);
+    }
+
+    private void Cancelled(DoAfterComponent component, DoAfter doAfter)
+    {
+        if (!component.DoAfters.TryGetValue(doAfter.ID, out var index))
+            return;
+
+        component.DoAfters.Remove(doAfter.ID);
+
+        if (component.DoAfters.Count == 0)
+            RemComp<ActiveDoAfterComponent>(component.Owner);
+
+        RaiseNetworkEvent(new CancelledDoAfterMessage(component.Owner, index.ID));
+    }
+
+    /// <summary>
+    ///     Call when the particular DoAfter is finished.
+    ///     Client should be tracking this independently.
+    /// </summary>
+    private void Finished(DoAfterComponent component, DoAfter doAfter)
+    {
+        if (!component.DoAfters.ContainsKey(doAfter.ID))
+            return;
+
+        component.DoAfters.Remove(doAfter.ID);
+
+        if (component.DoAfters.Count == 0)
+            RemComp<ActiveDoAfterComponent>(component.Owner);
+    }
+
+    private void OnStateChanged(EntityUid uid, DoAfterComponent component, MobStateChangedEvent args)
+    {
+        if (args.NewMobState != MobState.Dead || args.NewMobState != MobState.Critical)
+            return;
+
+        foreach (var (_, doAfter) in component.DoAfters)
+        {
+            Cancel(uid, doAfter, component);
+        }
+    }
+
+    /// <summary>
+    /// Cancels DoAfter if it breaks on damage and it meets the threshold
+    /// </summary>
+    /// <param name="uid">The EntityUID of the user</param>
+    /// <param name="component"></param>
+    /// <param name="args"></param>
+    private void OnDamage(EntityUid uid, DoAfterComponent component, DamageChangedEvent args)
+    {
+        if (!args.InterruptsDoAfters || !args.DamageIncreased || args.DamageDelta == null)
+            return;
+
+        foreach (var (_, doAfter) in component.DoAfters)
+        {
+            if (doAfter.EventArgs.BreakOnDamage && args.DamageDelta?.Total.Float() > doAfter.EventArgs.DamageThreshold)
+                Cancel(uid, doAfter, component);
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        foreach (var (_, comp) in EntityManager.EntityQuery<ActiveDoAfterComponent, DoAfterComponent>())
+        {
+            foreach (var (_, doAfter) in comp.DoAfters.ToArray())
```

Comment:
As above.

Reply:
Fixed the second one in https://github.com/space-wizards/space-station-14/pull/13225/commits/01f3c93afeb160420735aa628babce9e90b52fd9
The first one (EntityQuery) returns a tuple.